src/math/dct.c

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
+
+#include <sof/audio/format.h>
+#include <sof/math/matrix.h>
+#include <sof/math/dct.h>
+#include <sof/math/sqrt.h>
+#include <sof/math/trig.h>
+#include <errno.h>
+#include <stdint.h>
+
+#define PI_Q29 Q_CONVERT_FLOAT(3.1415926536, 29)
+#define TWO_PI_Q24 Q_CONVERT_FLOAT(6.2831853072, 24)
+#define ONE_OVER_SQRT_TWO Q_CONVERT_FLOAT(0.7071067812, 31)
+#define HALF_Q1 Q_CONVERT_FLOAT(0.5, 1)
+#define TWO_Q29 Q_CONVERT_FLOAT(2, 29)
+
+/* Return a matrix for discrete cosine transform (DCT). The DCT type is hard
+ * coded to DCT-II with orthogonal matrix. See
+ * https://en.wikipedia.org/wiki/Discrete_cosine_transform#DCT-II
+ *
+ * The size is (num_in, num_out). Currently there are not other input parameters.
+ * TODO: Add option for no ortho norm.
+ */
+
+/**
+ * \brief Initialize a 16 bit DCT matrix. The actual DCT transform is a matrix
+ * multiply with the returned matrix.
+ * \param[in,out]  dct  In input provide DCT type and size, in output the DCT matrix
+ */
+int dct_initialize_16(struct dct_plan_16 *dct)
+{
+	int16_t dct_val;
+	int32_t arg;
+	int32_t cos;
+	int32_t c1;
+	int16_t c2;
+	int16_t nk;
+	int n;
+	int m;
+	int k;
+
+	if (dct->type != DCT_II || dct->ortho != true)
+		return -EINVAL;
+
+	if (dct->num_in > DCT_MATRIX_SIZE_MAX || dct->num_out > DCT_MATRIX_SIZE_MAX)
+		return -EINVAL;
+
+	dct->matrix = mat_matrix_alloc_16b(dct->num_in, dct->num_out, 15);
+	if (!dct->matrix)
+		return -ENOMEM;
+
+	c1 = PI_Q29 / dct->num_in;
+	arg = Q_SHIFT_RND(TWO_Q29 / dct->num_in, 29, 12);
+	c2 = sqrt_int16(arg); /* Q4.12 */
+	for (n = 0; n < dct->num_in; n++) {
+		for (k = 0; k < dct->num_out; k++) {
+			nk = (Q_SHIFT_LEFT(n, 0, 1) + HALF_Q1) * Q_SHIFT_LEFT(k, 0, 1); /*Q14.2 */
+			arg = Q_MULTSR_32X32((int64_t)c1, nk, 29, 2, 24); /* Q8.24 */
```

Comment:
since nk is int16, I would suggest use 32*16, and keep result as 32 bit.
I am not sure how these comes, if possible, could you attach also with float point implementation?

Reply:
I'll check that. Currently the max size of DCT is limited by arg before unwrap. If DCT with 128 bins is needed need to change the Q-format. I'll do it if it doesn't weaken the precision. E.g. in Pytorch Kaldi the default is only 13 cepstral coefficients.

The float reference for DCT is here in second commit for unit test: https://github.com/thesofproject/sof/pull/6246/commits/59761996947670af17347bbe2710fa9e34dafcee#diff-49d5a9e56cdee8004d63f5442ccc4e4d1d24b9513ce425963d566e7540d69fb3

Reply:
With DCT_MATRIX_SIZE_MAX=42 the largest possible value of arg is 127.27. A Q8.24 can hold up to 127.99999 so this is safe also. I'll add a comment here too.

I need to revisit this if 128 or 256 long DCT is needed. Right now I'm more worried about fixed point accuracy vs. double and float reference MFCCs with small size DCT used. Precision is now more an issue than limitation of DCT size.

Reply:
Hi, Seppo, did you ever consider the fast DCT: wrapper / fft / unwrapper? or it is already fast dsp?